pkgs/cli_util/lib/src/base_directories.dart

```diff
@@ -0,0 +1,273 @@
+
+  /// Constructs a [BaseDirectories] instance for the given [tool] name.
+  BaseDirectories({
+    required this.tool,
+    Map<String, String>? environment,
+  }) : environment = environment ?? Platform.environment;
+
+  /// The path of the directory where [tool] can place its caches.
+  ///
+  /// The cache might at any point be purged by the operating system or user.
+  /// Applications should be able to reconstruct any data stored here. If [tool]
+  /// cannot handle data being purged, use [runtimeHome] or [dataHome] instead.
+  ///
+  /// This is a location appropriate for storing non-essential files that may be
+  /// removed at any point.
+  ///
+  /// The directory location depends on the current [Platform.operatingSystem]:
+  /// - on **Windows**:
+  ///   - `%LOCALAPPDATA%\<tool>`
+  /// - on **Mac OS**:
+  ///   - `$HOME/Library/Caches/<tool>`
+  /// - on **Linux**:
+  ///   - `$XDG_CACHE_HOME/<tool>` if `$XDG_CACHE_HOME` is defined, and
+  ///   - `$HOME/.cache/<tool>` otherwise.
+  ///
+  /// The directory won't be created, the method merely returns the recommended
+  /// location.
+  ///
+  /// On some platforms, this path may overlap with [runtimeHome] and
+  /// [stateHome].
+  ///
+  /// Throws an [EnvironmentNotFoundException] if the necessary environment
+  /// variables are undefined.
+  String cacheHome() =>
+      path.join(_baseDirectory(_XdgBaseDirectoryType.cache), tool);
+
+  /// The path of the directory where [tool] can place its configuration.
+  ///
+  /// The configuration may be synchronized across devices by the OS and may
+  /// survive application removal.
+  ///
+  /// This is a location appropriate for storing application specific
+  /// configuration for the current user.
+  ///
+  /// The directory location depends on the current [Platform.operatingSystem]
+  /// and what file types are stored:
+  /// - on **Windows**:
+  ///   - `%APPDATA%\<tool>`
+  /// - on **Mac OS**:
+  ///   - `$HOME/Library/Preferences/<tool>` (may only contain `.plist` files)
+  ///     if [plistFiles] is true, and
+  ///   - `$HOME/Library/Application Support/<tool>` otherwise.
+  /// - on **Linux**:
+  ///   - `$XDG_CONFIG_HOME/<tool>` if `$XDG_CONFIG_HOME` is defined, and
+  ///   - `$HOME/.config/<tool>` otherwise.
+  ///
+  /// The directory won't be created, the method merely returns the recommended
+  /// location.
+  ///
+  /// On some platforms, this path may overlap with [dataHome].
+  ///
+  /// Throws an [EnvironmentNotFoundException] if the necessary environment
+  /// variables are undefined.
+  String configHome({bool plistFiles = false}) {
+    if (Platform.isMacOS && !plistFiles) {
+      return dataHome();
+    }
+    return path.join(_baseDirectory(_XdgBaseDirectoryType.config), tool);
+  }
+
+  /// The path of the directory where [tool] can place its user data.
+  ///
+  /// The data may be backed up and synchronized to other devices by the
+  /// operating system. For large data use [stateHome] instead.
```

Comment:
I still don't think it's a "data size" thing on linux (which is the only platform where this and `stateHome` differ). Perhaps this is a Linux blunder and shouldn't be exposed to the developer.

I think [this wiki](https://wiki.gentoo.org/wiki/XDG/Base_Directories) says the difference most concisely, which might be good to add to this comment

> the base directory relative to which user-specific data files should be stored

for `stateHome`

> the [..] directory relative to which user-specific state files should be stored, containing data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in [`stateHome`]